Mathlib/Data/Polynomial/RingDivision.lean

```diff
@@ -854,6 +854,37 @@ theorem mem_nthRootsFinset {n : ℕ} (h : 0 < n) {x : R} :
 theorem nthRootsFinset_zero : nthRootsFinset 0 R = ∅ := by classical simp [nthRootsFinset_def]
 #align polynomial.nth_roots_finset_zero Polynomial.nthRootsFinset_zero
 
+theorem mul_mem_nthRootsFinset
+    {η₁ η₂ : R} (hη₁ : η₁ ∈ nthRootsFinset n R) (hη₂ : η₂ ∈ nthRootsFinset n R) :
+    η₁ * η₂ ∈ nthRootsFinset n R := by
+  cases n with
+  | zero =>
+    simp only [Nat.zero_eq, nthRootsFinset_zero, not_mem_empty] at hη₁
+  | succ n =>
+    rw [mem_nthRootsFinset n.succ_pos] at hη₁ hη₂ ⊢
+    rw [mul_pow, hη₁, hη₂, one_mul]
+
+theorem ne_zero_of_mem_nthRootsFinset {η : R} (hη : η ∈ nthRootsFinset n R) : η ≠ 0 := by
+  nontriviality R
+  rintro rfl
+  cases n with
+  | zero =>
+    simp only [Nat.zero_eq, nthRootsFinset_zero, not_mem_empty] at hη
+  | succ n =>
+    rw [mem_nthRootsFinset n.succ_pos, zero_pow n.succ_pos] at hη
+    exact zero_ne_one hη
+
+theorem one_mem_nthRootsFinset (hn : 0 < n) : 1 ∈ nthRootsFinset n R := by
+  rw [mem_nthRootsFinset hn, one_pow]
+
+def nthRootsSubsemigroup (n : ℕ) (R : Type*) [CommRing R] [IsDomain R] : Subsemigroup R where
```

Comment:
@riccardobrasca was this the kind of thing you were thinking of?

Reply:
I meant to change the definition of `nthRootsFinset` to make it a `Subsemigroup` instead of a `Finset`, so `mul_mem_nthRootsFinset` will be automatic. It's possible that this breaks a lot of things, in that case I don't think it is worth.

Reply:
Pushed - not very excited so far. I'd like to see the impact on flt-regular before landing this.

Reply:
Thanks. It should be possible to also add an instance of `Submonoid` provided we have `NeZero n`, adding an `example` that checks we are not creating a diamond.

Reply:
wip at https://github.com/leanprover-community/flt-regular/compare/mathlib4...nthrootfinset?expand=1